applications/RANSApplication/custom_conditions/vms_monolithic_k_based_wall_condition.cpp

```diff
@@ -0,0 +1,200 @@
+
+        mWallHeight = RansCalculationUtilities::CalculateWallHeight(*this, r_normal);
+
+        KRATOS_ERROR_IF(mWallHeight == 0.0) << this->Info() << " has zero wall height.\n";
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+std::string VMSMonolithicKBasedWallCondition<TDim, TNumNodes>::Info() const
+{
+    std::stringstream buffer;
+    buffer << "VMSMonolithicKBasedWallCondition" << TDim << "D";
+    return buffer.str();
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void VMSMonolithicKBasedWallCondition<TDim, TNumNodes>::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << "VMSMonolithicKBasedWallCondition";
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void VMSMonolithicKBasedWallCondition<TDim, TNumNodes>::PrintData(std::ostream& rOStream) const
+{
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void VMSMonolithicKBasedWallCondition<TDim, TNumNodes>::ApplyWallLaw(
+    MatrixType& rLocalMatrix,
+    VectorType& rLocalVector,
+    ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    if (RansCalculationUtilities::IsWallFunctionActive(*this)) {
+        const auto& r_geometry = this->GetGeometry();
+        // Get Shape function data
+        Vector gauss_weights;
+        Matrix shape_functions;
+        RansCalculationUtilities::CalculateConditionGeometryData(
+            r_geometry, this->GetIntegrationMethod(), gauss_weights, shape_functions);
+        const IndexType num_gauss_points = gauss_weights.size();
+
+        const size_t block_size = TDim + 1;
+
+        const double c_mu_25 = std::pow(rCurrentProcessInfo[TURBULENCE_RANS_C_MU], 0.25);
+        const double kappa = rCurrentProcessInfo[WALL_VON_KARMAN];
+        const double inv_kappa = 1.0 / kappa;
+        const double beta = rCurrentProcessInfo[WALL_SMOOTHNESS_BETA];
+        const double y_plus_limit = rCurrentProcessInfo[RANS_LINEAR_LOG_LAW_Y_PLUS_LIMIT];
+
+        const double eps = std::numeric_limits<double>::epsilon();
+
+        double condition_y_plus{0.0};
+        array_1d<double, 3> condition_u_tau = ZeroVector(3);
+
+        for (size_t g = 0; g < num_gauss_points; ++g) {
+            const Vector& gauss_shape_functions = row(shape_functions, g);
+
+            const array_1d<double, 3>& r_wall_velocity =
+                RansCalculationUtilities::EvaluateInPoint(
+                    r_geometry, VELOCITY, gauss_shape_functions);
+            const double wall_velocity_magnitude = norm_2(r_wall_velocity);
+
+            const double tke = RansCalculationUtilities::EvaluateInPoint(
+                r_geometry, TURBULENT_KINETIC_ENERGY, gauss_shape_functions);
+            const double rho = RansCalculationUtilities::EvaluateInPoint(
+                r_geometry, DENSITY, gauss_shape_functions);
+            const double nu = RansCalculationUtilities::EvaluateInPoint(
+                r_geometry, KINEMATIC_VISCOSITY, gauss_shape_functions);
```

Comment:
If I'm not wrong, in here you are looping all the nodes for each one of the magnitudes, that is the four times the `EvaluateInPoint` method is called. In this case I think it would be worth it to directly do the interpolation in here to loop the nodes once.

Reply:
done through discussed Parameter Pack based EvaluateInPoint :)